cub/cub/detail/rfa.cuh

```diff
@@ -0,0 +1,950 @@
+    {
+      return;
+    }
+
+    if (is_zero_v(primary(0)))
+    {
+      for (int i = 0; i < FOLD; i++)
+      {
+        primary(i * incpriY) = x.primary(i * incpriX);
+        carry(i * inccarY)   = x.carry(i * inccarX);
+      }
+      return;
+    }
+
+    if (is_nan_inf_v(x.primary(0)) || is_nan_inf_v(primary(0)))
+    {
+      primary(0) += x.primary(0);
+      return;
+    }
+
+    const auto X_index = x.binned_index();
+    const auto Y_index = this->binned_index();
+    const auto shift   = Y_index - X_index;
+    if (shift > 0)
+    {
+      // shift Y upwards and add X to Y
+      _CCCL_PRAGMA_UNROLL_FULL()
+      for (int i = FOLD - 1; i >= 1; i--)
+      {
+        if (i < shift)
+        {
+          break;
+        }
+        primary(i * incpriY) =
+          x.primary(i * incpriX) + (primary((i - shift) * incpriY) - binned_bins(i - shift + Y_index));
+        carry(i * inccarY) = x.carry(i * inccarX) + carry((i - shift) * inccarY);
+      }
+      _CCCL_PRAGMA_UNROLL_FULL()
+      for (int i = 0; i < FOLD; i++)
+      {
+        if (i == shift)
+        {
+          break;
+        }
+        primary(i * incpriY) = x.primary(i * incpriX);
+        carry(i * inccarY)   = x.carry(i * inccarX);
+      }
+    }
+    else if (shift < 0)
+    {
+      // shift X upwards and add X to Y
+      _CCCL_PRAGMA_UNROLL_FULL()
+      for (int i = 0; i < FOLD; i++)
+      {
+        if (i < -shift)
+        {
+          continue;
+        }
+        primary(i * incpriY) += x.primary((i + shift) * incpriX) - binned_bins(X_index + i + shift);
+        carry(i * inccarY) += x.carry((i + shift) * inccarX);
+      }
+    }
+    else if (shift == 0)
+    {
+      // add X to Y
+      _CCCL_PRAGMA_UNROLL_FULL()
+      for (int i = 0; i < FOLD; i++)
+      {
+        primary(i * incpriY) += x.primary(i * incpriX) - binned_bins(i + X_index);
+        carry(i * inccarY) += x.carry(i * inccarX);
+      }
+    }
+
+    binned_dmrenorm(incpriY, inccarY);
+  }
+
+  /// Add two manually specified binned fp (Y += X)
+  /// Performs the operation Y += X
+  _CCCL_DEVICE void binned_dbdbadd(const ReproducibleFloatingAccumulator& other)
+  {
+    binned_dmdmadd(other, 1, 1, 1, 1);
+  }
+
+public:
+  ReproducibleFloatingAccumulator()                                       = default;
+  ReproducibleFloatingAccumulator(const ReproducibleFloatingAccumulator&) = default;
+  /// Sets this binned fp equal to another binned fp
+  ReproducibleFloatingAccumulator& operator=(const ReproducibleFloatingAccumulator&) = default;
+
+  /// Set the binned fp to zero
+  _CCCL_DEVICE void zero()
+  {
+    data = {0};
+  }
+
+  /// Return the endurance of the binned fp
+  _CCCL_DEVICE _CCCL_FORCEINLINE constexpr int endurance() const
+  {
+    return ENDURANCE;
+  }
+
+  //! Accumulate an arithmetic @p x into the binned fp.
+  //! NOTE: Casts @p x to the type of the binned fp
+  template <typename U, typename ::cuda::std::enable_if_t<::cuda::std::is_arithmetic_v<U>>* = nullptr>
```

Comment:
Nitpick: We nowadays have our nice concept emulation available everywhere
```suggestion
  _CCCL_TEMPLATE(typename U)
  _CCCL_REQUIRES(::cuda::std::is_arithmetic_v<U>)
```

I believe that is much more readable than the old SFINAE. @bernhardmgruber are there any obejctions to using that throughout this PR?

Reply:
No objections! I tend to forget about it though :)